File: muxdb-py/muxdb/security/tls.py
from __future__ import annotations

import ssl
from typing import Optional

def create_client_ssl_context(
    ca_cert_path: Optional[str] = None,
    cert_chain_path: Optional[str] = None,
    private_key_path: Optional[str] = None,
) -> ssl.SSLContext:
    """Create an SSLContext for a secure client connection (e.g. gRPC or DB).

    Optionally configures mTLS if cert_chain_path and private_key_path are provided.
    """
    context = ssl.create_default_context(ssl.Purpose.SERVER_AUTH)
    
    if ca_cert_path:
        context.load_verify_locations(ca_cert_path)
    
    if cert_chain_path and private_key_path:
        context.load_cert_chain(certfile=cert_chain_path, keyfile=private_key_path)
        
    return context

def create_server_ssl_context(
    cert_chain_path: str,
    private_key_path: str,
    ca_cert_path: Optional[str] = None,
    require_client_auth: bool = False,
) -> ssl.SSLContext:
    """Create an SSLContext for a secure server.

    If require_client_auth is True, mTLS is enforced and ca_cert_path must be provided.
    """
    context = ssl.create_default_context(ssl.Purpose.CLIENT_AUTH)
    context.load_cert_chain(certfile=cert_chain_path, keyfile=private_key_path)
    
    if require_client_auth:
        if not ca_cert_path:
            raise ValueError("ca_cert_path is required when require_client_auth is True")
        context.load_verify_locations(ca_cert_path)
        context.verify_mode = ssl.CERT_REQUIRED
        
    return context

File: muxdb-js/src/security/tls.ts
import * as tls from 'tls';
import * as fs from 'fs';

export function createClientSSLOptions(
  caCertPath?: string,
  certChainPath?: string,
  privateKeyPath?: string
): tls.ConnectionOptions {
  /** Create standard Node.js ConnectionOptions for secure clients. */
  const options: tls.ConnectionOptions = {
    rejectUnauthorized: true,
  };
  
  if (caCertPath) {
    options.ca = fs.readFileSync(caCertPath);
  }
  
  if (certChainPath && privateKeyPath) {
    options.cert = fs.readFileSync(certChainPath);
    options.key = fs.readFileSync(privateKeyPath);
  }
  
  return options;
}

export function createServerSSLOptions(
  certChainPath: string,
  privateKeyPath: string,
  caCertPath?: string,
  requireClientAuth: boolean = false
): tls.SecureContextOptions & { requestCert?: boolean; rejectUnauthorized?: boolean } {
  /** Create SecureContextOptions for a Node.js server (e.g. gRPC or HTTPS). */
  const options: tls.SecureContextOptions & { requestCert?: boolean; rejectUnauthorized?: boolean } = {
    cert: fs.readFileSync(certChainPath),
    key: fs.readFileSync(privateKeyPath),
  };
  
  if (caCertPath) {
    options.ca = fs.readFileSync(caCertPath);
  }
  
  if (requireClientAuth) {
    if (!caCertPath) {
      throw new Error('caCertPath is required when requireClientAuth is true');
    }
    options.requestCert = true;
    options.rejectUnauthorized = true;
  }
  
  return options;
}

